spring-integration-file/src/main/java/org/springframework/integration/file/HeadDirectoryScanner.java

```diff
@@ -20,20 +20,40 @@
 import java.util.Arrays;
 import java.util.List;
 
+import org.springframework.integration.file.filters.CompositeFileListFilter;
 import org.springframework.integration.file.filters.FileListFilter;
 
 /**
  * A custom scanner that only returns the first <code>maxNumberOfFiles</code>
  * elements from a directory listing. This is useful to limit the number of File
- * objects in memory and therefore mutually exclusive with AcceptOnceFileListFilter.
- * 
+ * objects in memory and therefore mutually exclusive with {@code AcceptOnceFileListFilter}.
+ * It should not be used in conjunction with an {@code AcceptOnceFileListFilter}.
+ *
  * @author Iwein Fuld
+ * @author Gary Russell
  * @since 2.0
  */
 public class HeadDirectoryScanner extends DefaultDirectoryScanner {
 
+	private final HeadFilter headFilter;
+
 	public HeadDirectoryScanner(int maxNumberOfFiles) {
-		this.setFilter(new HeadFilter(maxNumberOfFiles));
+		HeadFilter headFilter = new HeadFilter(maxNumberOfFiles);
+		this.setFilter(headFilter);
+		this.headFilter = headFilter;
+	}
+
+	@Override
+	public void setFilter(FileListFilter<File> filter) {
+		if (filter instanceof CompositeFileListFilter) {
+			((CompositeFileListFilter<File>) filter).addFilter(this.headFilter);
+			super.setFilter(filter);
+		}
+		else {
+			CompositeFileListFilter<File> compositeFilter = new CompositeFileListFilter<File>();
+			compositeFilter.addFilter(filter).addFilter(this.headFilter);
```

Comment:
It actually doesn't make a difference; I thought the `CompositeFLF` works that way, but in fact the original list is presented to every filter and only those files that pass every filter are passed by the CompositeFilter. It doesn't matter what order the filters are invoked.